# IIC2413 - Pauta Control 2: SQL e Índices

## Parte 1: Consultas SQL (5 puntos) 

Las soluciones a cada pregunta son solo de ejemplo ya que hay varias formas de escribir las consultas que se piden.
Cada consulta vale lo que se especifica en ella y otorgar puntajes intermedios queda a criterio del corrector.

### Setup

In [ ]:
# Ejecuta esta celda para descargar la base de datos a colab
!wget  https://github.com/IIC2413/Syllabus-2021-1/raw/main/Controles/Control%202/tu_shop_mini.db

--2021-05-28 20:35:09--  https://github.com/IIC2413/Syllabus-2021-1/raw/main/Controles/Control%202/tu_shop_mini.db
Resolving github.com (github.com)... 140.82.114.3
Connecting to github.com (github.com)|140.82.114.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/IIC2413/Syllabus-2021-1/main/Controles/Control%202/tu_shop_mini.db [following]
--2021-05-28 20:35:09--  https://raw.githubusercontent.com/IIC2413/Syllabus-2021-1/main/Controles/Control%202/tu_shop_mini.db
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 819200 (800K) [application/octet-stream]
Saving to: ‘tu_shop_mini.db’

tu_shop_mini.db     100%[===================>] 800.00K  --.-KB/s    in 0.09s   

2021-05-28 20:35:10 (8.84 MB/s) - 

In [ ]:
%load_ext sql

In [ ]:
%sql sqlite:///tu_shop_mini.db

'Connected: @tu_shop_mini.db'

In [ ]:
%sql SELECT COUNT(*) FROM Bills

 * sqlite:///tu_shop_mini.db
Done.


[(2767,)]

Si la respuesta a la consulta de arriba es 2767 entonces estás listo para ejecutar las consultas de la pauta.

# PAUTA:  General Parte Consultas

- Descuento general de **0.1** si es que no muestran en el resultado algo pedido explícitamente en el enunciado (ej: si se pide nombre y fecha, es mínimo que estén esos dos, si muestran solo nombre, descuentan 0.1, si muestran solo id, así que no muestran nombre ni fecha descuentan 0.1 tmbn. Da lo mismo si muestran más de lo pedido, onda muestran id nombre y fecha, tienen todo el puntaje)

### Consultas

**1.** El id y nombre de cada producto junto con el precio y cantidad de su master variant  ordenados ascendente por 'row_order'. **(0.3 pts)**

Pauta: 
- 0.1 por hacer el join entre productos y variantes
- 0.1 por que sean productos (el is_master true)
- 0.1 por ordenar ascendente

In [ ]:
%%sql
SELECT Products.id, Products.name, Variants.price, Variants.quantity
FROM Products JOIN Variants
ON Products.id = Variants.product_id
WHERE Variants.is_master='true'
ORDER BY Products.row_order

 * sqlite:///tu_shop_mini.db
Done.


[(42766, 'Post Scriptum', 34990, 0),
 (42339, 'Pummel Party', 1, 1),
 (42717, 'Diablo III', 1, 1),
 (6540, 'Assassin’s Creed® Unity', 1, 1),
 (42240, 'Pummel Party', 1, 1),
 (42716, 'Crying Suns', 1, 1),
 (6538, 'Argo', 1, 1),
 (12718, 'Gloomhaven', 34900, 10),
 (42715, 'Call of Duty®: Black Ops 4', 1, 1),
 (726, 'Batman: Arkham City', 1, 1),
 (6536, 'Ice Climbers', 1, 1),
 (42714, 'Darksiders II Deathinitive Edition', 1, 1),
 (3069, 'EVERSPACE™', 58900, 0),
 (6535, 'Kid Icarus', 1, 1),
 (42395, 'The Evil Within 2', 1, 1),
 (724, 'God of War', 1, 1),
 (6534, 'L.A. Noire', 1, 1),
 (42394, 'The Crew™', 1, 1),
 (609, 'Left 4 Dead', 1, 1),
 (6533, 'Yoshis Wooly World', 1, 1),
 (37143, 'Darkest Dungeon®', 1, 1),
 (723, 'Dishonored', 1, 1),
 (42393, 'The Crew® 2', 1, 1),
 (3026, 'BioShock™ Remastered', 42000, 20),
 (6156, 'Halo 5: Guardians', 1, 1),
 (6532, 'Anno 1800™', 1, 1),
 (41759, 'Pokémon: Lets Go Pikachu!', 1, 1),
 (476, 'Civilization VI', 20000, 0),
 (1432, 'Call of Duty®: Modern Warfare®', 8990, 3),
 (20382, 'Praetorians - HD Remaster', 1, 1),
 (24604, 'Tabletop Simulator', 16990, 7),
 (41758, 'Wii Sports', 1, 1),
 (20187, 'Pillars of Eternity II: Deadfire', 1, 1),
 (6530, 'Anno 1800™', 1, 1),
 (24571, 'StarCraft II®', 1, 1),
 (255, 'Team Fortress 2', 1, 1),
 (719, 'Super Meat Boy', 9500, 4),
 (41757, 'Animal Crossing: City Folk', 1, 1),
 (20186, 'XERA: Survival', 1, 1),
 (473, 'Transistor', 1, 1),
 (5290, 'Half-Life 2: Lost Coast', 1, 1),
 (6529, 'Kirbys Epic Yarn', 1, 1),
 (18735, 'StarCraft II®', 1, 1),
 (5358, 'Pokémon SoulSilver', 1, 1),
 (41756, 'Hearthstone', 1, 1),
 (472, 'Animal Crossing: Wild World', 1, 1),
 (3022, 'Batman™: Arkham City - Game of the Year Edition', 42000, 20),
 (6528, 'Kirbys Dream Land', 1, 1),
 (4791, 'Assassin’s Creed®: Syndicate', 1, 1),
 (41755, 'Overwatch', 1, 1),
 (43009, 'Counter-Strike: Global Offensive', 1, 1),
 (1121, 'Civilization IV', 10990, 1),
 (42774, 'Project CARS 2', 1, 1),
 (9332, 'Crysis 2', 10990, 8),
 (12833, "Garry's Mod", 1, 1),
 (6527, 'Albion Online', 1, 1),
 (3670, "LEGO® MARVEL's Avengers", 24990, 0),
 (41754, 'Golf With Your Friends', 1, 1),
 (42741, 'Lonely Mountains: Downhill', 12000, 0),
 (1426, 'Super Mario Bros. 3', 8990, 3),
 (4788, 'ARK: Survival Of The Fittest', 1, 1),
 (42773, 'Project Winter', 1, 1),
 (12832, "Garry's Mod", 1, 1),
 (6526, 'Albino Lullaby: Episode 1', 1, 1),
 (309, 'The Legend of Zelda: Twilight Princess', 10990, 2),
 (3669, "Leisure Suit Larry - Wet Dreams Don't Dry", 24990, 0),
 (39029, 'Assetto Corsa Competizione', 1, 1),
 (42740, 'Lethal League Blaze', 12000, 0),
 (42771, 'PUBG: Closed Experimental Server', 1, 1),
 (12831, 'Guild Wars 2', 1, 1),
 (1479, 'AO International Tennis', 24490, 1),
 (3668, 'LEGO® Marvel™ Super Heroes', 24990, 0),
 (39028, 'Just Cause 2', 1, 1),
 (42739, 'LEGO® Marvel Super Heroes 2', 1, 1),
 (39891, 'Spyro™ Reignited Trilogy', 42800, 0),
 (42770, 'Project CARS', 1, 1),
 (12830, 'GUILTY GEAR Xrd -REVELATOR-', 1, 1),
 (42936, "Tom Clancy's Ghost Recon® Wildlands", 38900, 3),
 (6524, 'Kirbys Adventure', 1, 1),
 (307, 'Civilization IV', 10990, 0),
 (42811, 'Star Control®: Origins', 10990, 8),
 (11378, 'Call of Duty: Modern Warfare Remastered', 1, 1),
 (912, 'Halo: Combat Evolved', 38900, 1),
 (1473, 'Dwarf Fortress', 1, 1),
 (39027, 'Batman: Arkham Asylum', 1, 1),
 (42644, "Yoku's Island Express", 14900, 2),
 (6927, 'Last Tide', 48900, 0),
 (42769, 'Project CARS', 1, 1),
 (42795, 'Reigns: Game of Thrones', 1, 1),
 (12829, 'Golf With Your Friends', 1, 1),
 (467, 'Halo 4', 1, 1),
 (6523, '7 Days to Die', 1, 1),
 (42810, 'Pokémon Ruby', 11990, 7),
 (39023, 'Friday the 13th: The Game', 1, 1),
 (42643, 'Yakuza Kiwami', 14900, 1),
 (11377, 'The Witcher 3: Wild Hunt', 1, 1),
 (42639, 'Startup Company', 25990, 1),
 (42804, 'RIDE 3', 9500, 10),
 (42797, 'Remnant: From the Ashes', 7990, 6),
 (466, 'Pokémon Y', 1, 1),
 (551, 'Half-Life 2', 9500, 1),
 (6522, 'PlayerUnknowns Battlegrounds', 1, 1),
 (42809, 'Super Mario Bros. 2', 

**2.** El status y la fecha de creacion de las ventas de la tienda `tío zetamart` en marzo del 2021, ordenados por la fecha de creacion. **(0.3 pts)**

Pauta: 
- 0.1 por condición de fecha
- 01. por condición de nombre de la tienda (tenía que ser sobre el nombre, si buscaron el id y luego lo usaron aquí, descontar este punto)
- 0.1 por ordenar (debería ser asecendente, pero no lo dice explícito el enunciado así que con esté ordenado de alguna forma, esta bien)

In [ ]:
%%sql
SELECT Bills.status, Bills.created_at
FROM Bills JOIN Stores
ON Bills.store_id = Stores.id
WHERE Stores.name ='tío zetamart'
AND Bills.created_at BETWEEN '2021-03-01' AND '2021-04-01'
ORDER BY Bills.created_at

 * sqlite:///tu_shop_mini.db
Done.


[('payment_accepted', '2021-03-01 21:46:34 -0300'),
 ('deleted', '2021-03-02 02:21:42 -0300'),
 ('payment_accepted', '2021-03-02 15:30:20 -0300'),
 ('payment_accepted', '2021-03-02 18:40:35 -0300'),
 ('deleted', '2021-03-02 19:39:13 -0300'),
 ('payment_accepted', '2021-03-02 20:11:37 -0300'),
 ('deleted', '2021-03-02 21:22:05 -0300'),
 ('payment_accepted', '2021-03-02 21:35:41 -0300'),
 ('payment_accepted', '2021-03-03 11:20:33 -0300'),
 ('deleted', '2021-03-03 12:45:30 -0300'),
 ('deleted', '2021-03-03 13:54:26 -0300'),
 ('payment_accepted', '2021-03-03 14:00:54 -0300'),
 ('payment_accepted', '2021-03-04 10:51:48 -0300'),
 ('deleted', '2021-03-04 19:15:48 -0300'),
 ('deleted', '2021-03-04 23:05:34 -0300'),
 ('payment_accepted', '2021-03-05 15:43:50 -0300'),
 ('payment_accepted', '2021-03-06 12:52:41 -0300'),
 ('payment_accepted', '2021-03-07 13:06:56 -0300'),
 ('payment_accepted', '2021-03-07 19:31:04 -0300'),
 ('payment_accepted', '2021-03-07 21:14:46 -0300'),
 ('payment_accepted', '2021-03-08 12:07:54 -0300'),
 ('deleted', '2021-03-09 11:41:16 -0300'),
 ('deleted', '2021-03-09 23:33:18 -0300'),
 ('payment_accepted', '2021-03-09 23:42:15 -0300'),
 ('shipped', '2021-03-09 23:42:16 -0300'),
 ('payment_accepted', '2021-03-10 19:03:31 -0300'),
 ('deleted', '2021-03-10 19:15:10 -0300'),
 ('deleted', '2021-03-10 19:48:17 -0300'),
 ('payment_accepted', '2021-03-10 19:51:18 -0300'),
 ('payment_accepted', '2021-03-11 08:50:24 -0300'),
 ('deleted', '2021-03-11 12:51:34 -0300'),
 ('payment_accepted', '2021-03-11 14:27:54 -0300'),
 ('deleted', '2021-03-11 14:35:17 -0300'),
 ('payment_accepted', '2021-03-11 18:45:41 -0300'),
 ('payment_accepted', '2021-03-11 20:56:50 -0300'),
 ('payment_accepted', '2021-03-11 21:36:17 -0300'),
 ('payment_accepted', '2021-03-12 05:10:16 -0300'),
 ('payment_accepted', '2021-03-12 08:31:37 -0300'),
 ('payment_accepted', '2021-03-12 15:49:48 -0300'),
 ('payment_accepted', '2021-03-14 10:19:27 -0300'),
 ('payment_accepted', '2021-03-14 14:05:02 -0300'),
 ('awaiting_payment', '2021-03-15 17:41:03 -0300'),
 ('shipped', '2021-03-16 15:27:11 -0300'),
 ('deleted', '2021-03-17 22:15:25 -0300'),
 ('payment_accepted', '2021-03-18 23:27:29 -0300'),
 ('deleted', '2021-03-19 11:22:10 -0300'),
 ('payment_accepted', '2021-03-19 12:24:43 -0300'),
 ('deleted', '2021-03-19 12:26:10 -0300'),
 ('payment_accepted', '2021-03-19 13:23:39 -0300'),
 ('payment_accepted', '2021-03-19 15:37:04 -0300'),
 ('payment_accepted', '2021-03-19 15:48:07 -0300'),
 ('awaiting_payment', '2021-03-19 21:27:44 -0300'),
 ('payment_accepted', '2021-03-20 18:31:00 -0300'),
 ('deleted', '2021-03-20 20:39:02 -0300'),
 ('deleted', '2021-03-20 22:47:35 -0300'),
 ('deleted', '2021-03-21 00:09:06 -0300'),
 ('payment_accepted', '2021-03-21 00:11:43 -0300'),
 ('payment_accepted', '2021-03-21 09:16:52 -0300'),
 ('deleted', '2021-03-21 11:43:41 -0300'),
 ('payment_accepted', '2021-03-21 12:34:58 -0300'),
 ('payment_accepted', '2021-03-21 12:56:45 -0300'),
 ('payment_accepted', '2021-03-21 15:33:19 -0300'),
 ('payment_accepted', '2021-03-21 17:39:15 -0300'),
 ('payment_accepted', '2021-03-21 18:30:06 -0300'),
 ('payment_accepted', '2021-03-21 19:56:17 -0300'),
 ('shipped', '2021-03-21 20:05:51 -0300'),
 ('deleted', '2021-03-21 20:13:22 -0300'),
 ('awaiting_payment', '2021-03-24 22:47:36 -0300'),
 ('deleted', '2021-03-25 15:08:45 -0300'),
 ('received', '2021-03-25 15:20:10 -0300'),
 ('payment_accepted', '2021-03-29 21:20:36 -0300'),
 ('payment_accepted', '2021-03-29 21:52:54 -0300'),
 ('payment_accepted', '2021-03-29 21:56:50 -0300'),
 ('payment_accepted', '2021-03-29 21:58:20 -0300'),
 ('shipped', '2021-03-29 22:44:29 -0300'),
 ('deleted', '2021-03-29 23:44:17 -0300'),
 ('deleted', '2021-03-30 01:07:16 -0300'),
 ('payment_accepted', '2021-03-30 09:51:38 -0300'),
 ('payment_accepted', '2021-03-30 12:07:14 -0300'),
 ('payment_accepted', '2021-03-30 13:33:17 -0300'),
 ('shipped', '2021-03-30 14:00:17 -0300')]

**3.** Los emails de las personas que hayan comprado al menos una vez en la tienda `pc_faketory` antes de abril del 2021. **(0.3 pts)**

Pauta: 
- 0.1 por hacer el join entre tiendas y productos
- 0.1 por condición de la tienda sobre el nombre (si buscó la tienda y puso el id, descontar este punto)
- 0.1 por condición de fecha

Da lo mismo si le pusieron `DISTINCT` o no

In [ ]:
%%sql
SELECT DISTINCT Bills.email
FROM Bills, Stores
WHERE Bills.store_id = Stores.id AND Stores.name = 'pc_faketory'
AND Bills.created_at < '2021-04-01'

 * sqlite:///tu_shop_mini.db
Done.


[('gonzalo33@rossello.com',),
 ('telmotejero@lumbreras.es',),
 ('colladolorenza@hotmail.com',),
 ('maximiano17@robles.es',),
 ('jquiros@barreda.com',),
 ('maite24@gmail.com',),
 ('cparra@hervas-escamilla.es',),
 ('umartin@saenz.com',),
 ('arrietafabiola@feijoo.com',),
 ('samuacero@villalba-perea.net',),
 ('maria-del-carmenpatino@gmail.com',),
 ('riosraul@yahoo.com',),
 ('omanrique@buendia.es',),
 ('zvilla@vendrell-pujadas.net',),
 ('zalvarez@hotmail.com',),
 ('gervasioaguilera@perera.com',),
 ('fabian27@isern.org',),
 ('maristelavalbuena@ferreras.com',),
 ('beldapepe@gmail.com',),
 ('gabaldonmaricruz@hotmail.com',),
 ('qsoler@gmail.com',),
 ('llanomilagros@gmail.com',),
 ('trinidad71@hotmail.com',),
 ('trosales@hotmail.com',),
 ('cheloraya@gmail.com',),
 ('reynatomas@hotmail.com',),
 ('jvendrell@bonet.com',),
 ('cobosursula@cerezo.es',),
 ('lucianamoya@yahoo.com',),
 ('carmelita23@yahoo.com',),
 ('umilla@bilbao.com',),
 ('lestrada@jover-clavero.com',),
 ('fguardia@barbero.com',),
 ('tamaramoles@bernal.com',),
 ('quiroscesar@hotmail.com',),
 ('pabloarevalo@olive.com',),
 ('luquelourdes@hotmail.com',),
 ('brionesadora@robles.es',),
 ('lvillalonga@cid.es',),
 ('andreavaquero@gmail.com',),
 ('julieta62@hotmail.com',),
 ('pintoloreto@armengol.net',),
 ('maria-dolores70@hotmail.com',),
 ('nacio63@portero-leal.com',),
 ('luzchita@bravo.net',),
 ('quintanillafelipe@yahoo.com',),
 ('juanitobuendia@gmail.com',),
 ('sarita00@gmail.com',),
 ('jordi97@gutierrez-conesa.com',),
 ('npont@yahoo.com',),
 ('desiderio89@arce.es',),
 ('benita49@yahoo.com',),
 ('mollignacia@rosell-gallego.es',),
 ('vizcainobenjamin@hotmail.com',),
 ('manuelitaparra@yahoo.com',),
 ('arsenio56@conde.com',),
 ('salemany@lara.org',),
 ('sblazquez@yahoo.com',),
 ('maria-fernanda30@yahoo.com',),
 ('saavedrajordana@hotmail.com',),
 ('torresvaleria@yahoo.com',),
 ('segismundo67@gmail.com',),
 ('paulaalegria@comas.org',),
 ('solaemiliana@yahoo.com',),
 ('rnicolas@moreno.es',),
 ('anabel87@amaya.es',),
 ('costabruno@hernando.es',),
 ('julian62@domingo.es',),
 ('jordipineiro@hotmail.com',),
 ('norbertoarnaiz@yahoo.com',),
 ('kborrell@yahoo.com',),
 ('carranzaleandro@vicente.es',),
 ('abilioclemente@aparicio.es',),
 ('ariel01@cerro-lucena.es',),
 ('heribertoclavero@gmail.com',),
 ('reynaldogras@montero.com',),
 ('felisacozar@berenguer.es',),
 ('sabina18@solana.es',),
 ('rosendapino@diego.com',),
 ('americo99@yahoo.com',),
 ('zibanez@boada-toledo.es',),
 ('ribaodalys@gmail.com',),
 ('andreafajardo@pina.org',),
 ('lmiro@alcaraz.es',),
 ('mcalzada@yahoo.com',),
 ('tudelaeli@yahoo.com',),
 ('ursula27@calzada.org',),
 ('ruy69@yahoo.com',),
 ('pjulia@yahoo.com',),
 ('eustaquionovoa@yahoo.com',),
 ('bonifaciolarrea@gmail.com',),
 ('iris37@yahoo.com',),
 ('aristides21@colom.net',),
 ('valverdebenita@tudela.com',),
 ('escolanoleocadio@codina.es',),
 ('lara48@yahoo.com',),
 ('eligia29@quesada.com',),
 ('bvalencia@franch-ripoll.com',),
 ('santanaagustin@roldan.es',),
 ('palmagabriela@gmail.com',),
 ('pujadasfrancisca@torrens-julia.es',),
 ('quintanaandrea@melendez.com',),
 ('ernestosegarra@hotmail.com',),
 ('cornejosarita@hotmail.com',),
 ('caparrosclara@morales-amaya.es',),
 ('miguelmohamed@yahoo.com',),
 ('primitivoromeu@gutierrez.org',),
 ('jeronimocuervo@yahoo.com',),
 ('bartolome59@hotmail.com',),
 ('jaimepadilla@andrade-adadia.org',),
 ('dblanch@segura.com',),
 ('ipuig@nogueira.com',),
 ('dorotea88@cabello-juan.es',),
 ('caridad89@gmail.com',),
 ('bbarroso@zamora.net',),
 ('melchor50@yahoo.com',),
 ('maria-carmeniniesta@yahoo.com',),
 ('spedrero@bueno.com',),
 ('ovara@luz.es',),
 ('tomassabina@caro.net',),
 ('paco92@vergara.com',),
 ('imelda08@bayona.com',),
 ('fannycastell@galiano.com',),
 ('avall@reig.com',),
 ('ariadnafrias@gmail.com',),
 ('zzorrilla@losada.es',),
 ('luciaroldan@pacheco.com',),
 ('herminio86@quiros.org',),
 ('joaquinafeliu@gmail.com',),
 ('gcarlos@belda.es',),
 ('gcoronado@gmail.com',),
 ('jferrando@gmail.com',),
 ('regulo78@valero.es',),
 ('falegre@hotmail.com'

**4.** Los nombres completos, precio y cantidad y valor total (precio * cantidad) de cada variante. **(0.3 pts)**

Pauta: 
- 0.1 por hacer el join entre productos y variantes
- 0.1 por desplegar el nombre completo (producto + sufijo)
- 0.1 por valor total como la multiplicación de precio y cantidad

In [ ]:
%%sql
SELECT (Products.name || ' ' || Variants.name_suffix) AS full_name, Variants.price, Variants.quantity, (Variants.price * Variants.quantity ) AS total_value
FROM Products JOIN Variants
ON Products.id = Variants.product_id

 * sqlite:///tu_shop_mini.db
Done.


[('Civilization IV ', 9990, 5, 49950),
 ('The Legend of Zelda: Breath of the Wild ', 9990, 4, 39960),
 ('Civilization IV ', 9990, 9, 89910),
 ('Civilization IV ', 9990, 5, 49950),
 ('System Shock ', 400, 1, 400),
 ('Civilization IV ', 9990, 15, 149850),
 ('The Legend of Zelda: Breath of the Wild ', 9990, 4, 39960),
 ('The Legend of Zelda: Majoras Mask ', 9990, 11, 109890),
 ('Day of Defeat ', 9990, 0, 0),
 ('The Legend of Zelda: Majoras Mask ', 9990, 4, 39960),
 ('Pong ', 9990, 4, 39960),
 ('Civilization III ', 10990, 6, 65940),
 ('Amazing Cultivation Simulator / 了不起的修仙模拟器 ', 9990, 2, 19980),
 ('Civilization III ', 9990, 2, 19980),
 ('Pokémon Ruby ', 9990, 2, 19980),
 ('Commandos 2: Men of Courage ', 9990, 4, 39960),
 ('Half-Life: Opposing Force ', 9990, 3, 29970),
 ('Call of Duty: Black Ops III ', 9990, 2, 19980),
 ('Civilization IV ', 9990, 0, 0),
 ('Portal ', 9990, 1, 9990),
 ('The Legend of Zelda: Majoras Mask ', 9990, 5, 49950),
 ('Pokémon Ruby ', 10990, 7, 76930),
 ('WarioWare: Touched! ', 10990, 10, 109900),
 ('The Legend of Zelda: Skyward Sword ', 10990, 1, 10990),
 ('Dota 2 ', 10990, 0, 0),
 ('SpellForce 3 ', 10990, 6, 65940),
 ('Portal 2 ', 10990, 5, 54950),
 ('Civilization IV ', 10990, 10, 109900),
 ('FTL: Faster Than Light ', 10990, 7, 76930),
 ('Super Mario Sunshine ', 9990, 0, 0),
 ('Pokémon Ruby ', 9990, 3, 29970),
 ('The Legend of Zelda: Ocarina of Time ', 9990, 3, 29970),
 ('Civilization V ', 9990, 10, 99900),
 ('Commandos: Behind Enemy Lines ', 9990, 5, 49950),
 ('Pac-Man ', 10990, 2, 21980),
 ('The Legend of Zelda: Breath of the Wild ', 10990, 5, 54950),
 ('Civilization IV ', 10990, 7, 76930),
 ('Pong ', 10990, 11, 120890),
 ('Castle Crashers ', 10990, 2, 21980),
 ('Battlefleet Gothic: Armada 2 ', 11990, 3, 35970),
 ('Starcraft ', 10990, 1, 10990),
 ('Civilization IV ', 10990, 7, 76930),
 ('Quake ', 10990, 1, 10990),
 ('American Truck Simulator ', 10990, 8, 87920),
 ('Pokémon White ', 8990, 0, 0),
 ('Pokémon Black ', 8990, 0, 0),
 ('Pokémon Blue ', 8990, 2, 17980),
 ('Pokémon Black 2 ', 8990, 0, 0),
 ('Pokémon White 2 ', 6500, 0, 0),
 ('Life is Strange ', 8990, 0, 0),
 ('Civilization IV ', 10990, 4, 43960),
 ('Civilization IV ', 10990, 2, 21980),
 ('Left 4 Dead 2 ', 10990, 5, 54950),
 ('The Legend of Zelda: Breath of the Wild ', 10990, 7, 76930),
 ('Super Mario Bros. ', 10990, 5, 54950),
 ('Pokémon Alpha Sapphire ', 1, 1, 1),
 ('Pokémon Alpha Sapphire 6', 12990, 1, 12990),
 ('Pokémon Alpha Sapphire 7', 12990, 1, 12990),
 ('Pokémon Alpha Sapphire 8', 12990, 1, 12990),
 ('Pokémon Sun 6', 11990, 1, 11990),
 ('Pokémon Sun 8', 11990, 1, 11990),
 ('Pokémon Sun ', 1, 1, 1),
 ('Pokémon Moon 6', 11990, 1, 11990),
 ('Pokémon Moon 7', 11990, 1, 11990),
 ('Pokémon Moon ', 1, 1, 1),
 ('Pokémon LeafGreen 8', 14990, 10, 149900),
 ('Pokémon LeafGreen ', 1, 1, 1),
 ('Team Fortress 2 7', 13990, 10, 139900),
 ('Team Fortress 2 8', 13990, 10, 139900),
 ('Team Fortress 2 ', 1, 1, 1),
 ('Team Fortress 2 ', 1, 1, 1),
 ('Team Fortress 2 8', 10990, 10, 109900),
 ('Pokémon Sapphire 8', 11990, 10, 119900),
 ('Pokémon Sapphire ', 1, 1, 1),
 ('Team Fortress 2 8', 10990, 10, 109900),
 ('Team Fortress 2 ', 1, 1, 1),
 ('Civilization IV ', 10990, 7, 76930),
 ('Team Fortress 2 6', 11990, 10, 119900),
 ('Team Fortress 2 7', 11990, 10, 119900),
 ('Team Fortress 2 8', 11990, 10, 119900),
 ('Team Fortress 2 ', 1, 1, 1),
 ('Pokémon HeartGold 7', 11990, 10, 119900),
 ('Pokémon HeartGold 8', 11990, 11990, 143760100),
 ('Pokémon HeartGold ', 1, 1, 1),
 ('Pokémon Pearl ', 1, 1, 1),
 ('Pokémon Pearl 8', 11990, 10, 119900),
 ('Pokémon Platinum 7', 11990, 10, 119900),
 ('Pokémon Platinum 8', 11990, 10, 119900),
 ('Pokémon Platinum ', 1, 1, 1),
 ('Pokémon Yellow 6', 12990, 1, 12990),
 ('Pokémon Yellow 7', 12990, 1, 12990),
 ('Pokémon Yellow 8', 12990, 1, 12990),
 ('Pokémon Yellow ', 1, 1, 1),
 ('Half-Life: Blue Shift 7', 11990, 10, 119900),
 ('Half-Life: Blue Shift 8', 11990, 10, 119900),
 ('Half-Life: Blue Shift ', 1, 1, 1),
 ('Halo 3 6', 11990, 10, 119

**5.** El nombre de la tienda y la dirección de envío de las ventas que ya hayan sido enviadas (`shipped`) y tengan 'Cádiz' en la dirección. **(0.3 pts)**

Pauta: 
- 0.1 por hacer el join
- 0.2 porcondición sobre dirección usando LIKE

La respuesta debería con la condición de `shipped`, pero si no la pusieron, no descontar. Si bien está en el enunciado de la consulta, como en las issues dije que no consideraran el status (obviamente esto iba para las demás consultas), igual puede haber pasado que alguien tampoco la considerara acá.

In [ ]:
%%sql
SELECT Stores.name, Bills.address
FROM Bills JOIN Stores ON Bills.store_id = Stores.id
WHERE Bills.address LIKE '%Cádiz%'

 * sqlite:///tu_shop_mini.db
Done.


[('Kwik E-mart', 'Via de Ariel Piña 16 Cádiz, 50873'),
 ('pc_faketory', 'Camino de Jose Luis Crespi 76 Puerta 7  Cádiz, 18683'),
 ('Comic center', 'Vial de Matías Castell 90 Puerta 9  Cádiz, 91837'),
 ('Comic center', 'Vial de Matías Castell 90 Puerta 9  Cádiz, 91837'),
 ('tío zetamart', 'Cañada Federico Yáñez 45 Puerta 3  Cádiz, 74801'),
 ('Comic center', 'Vial Diana Olivé 55 Puerta 2  Cádiz, 75112'),
 ('Comic center', 'Paseo de Federico Cabezas 52 Cádiz, 49809'),
 ('tío zetamart', 'Callejón Juan José Cobo 21 Cádiz, 75367'),
 ('tío zetamart', 'Pasaje de Adora Quirós 83 Cádiz, 08989'),
 ('instatienda', 'Rambla Edelmira Bermudez 2 Puerta 9  Cádiz, 83315'),
 ('pc_faketory', 'Acceso de Paulina Adán 80 Cádiz, 79326'),
 ('instatienda', 'Alameda de María Carmen Miralles 4 Cádiz, 23077'),
 ('instatienda', 'Camino de Demetrio Cazorla 250 Puerta 7  Cádiz, 18940'),
 ('pc_faketory', 'Vial de Ofelia Valderrama 406 Cádiz, 02910'),
 ('pc_faketory', 'Callejón de Alma Ibarra 23 Piso 7  Cádiz, 19821'),
 ('instatienda', 'Paseo de Delfina Aguilar 17 Cádiz, 26136'),
 ('instatienda', 'Pasaje de Fabián Fabra 3 Apt. 09  Cádiz, 73863'),
 ('pc_faketory', 'Via de Fortunata Manzano 33 Puerta 3  Cádiz, 86233'),
 ('pc_faketory', 'Avenida Leire Durán 60 Puerta 0  Cádiz, 71001'),
 ('instatienda', 'Cuesta de Hernando Palacio 6 Apt. 73  Cádiz, 92657'),
 ('instatienda', 'Acceso Olivia Acevedo 766 Apt. 57  Cádiz, 67023'),
 ('pc_faketory', 'Cuesta Abel Sanmiguel 57 Piso 6  Cádiz, 50843'),
 ('pc_faketory', 'C. Néstor Aguado 345 Piso 2  Cádiz, 53783'),
 ('tío zetamart', 'C. Laura Toledo 309 Piso 7  Cádiz, 43161'),
 ('Comic center', 'Urbanización Ramona Mariscal 117 Cádiz, 75610'),
 ('instatienda', 'C. Elisabet Ferrán 87 Cádiz, 17486'),
 ('pc_faketory', 'Camino de Yéssica Cardona 33 Apt. 47  Cádiz, 36534'),
 ('pc_faketory', 'Avenida de Salvador Serna 598 Apt. 25  Cádiz, 60809'),
 ('instatienda', 'Urbanización Régulo Elías 6 Apt. 58  Cádiz, 40117'),
 ('pc_faketory', 'Pasadizo de Ramona Angulo 607 Piso 5  Cádiz, 30441'),
 ('instatienda', 'Acceso Rosaura Espinosa 20 Cádiz, 36106'),
 ('pc_faketory', 'Avenida María Manuela Canet 10 Cádiz, 56314'),
 ('pc_faketory', 'Rambla de Adelardo Vilalta 9 Cádiz, 93212'),
 ('pc_faketory', 'Camino Manuela Barros 31 Cádiz, 16967'),
 ('pc_faketory', 'Camino de Reina Sáenz 77 Cádiz, 30013'),
 ('tío zetamart', 'Cañada Federico Yáñez 45 Puerta 3  Cádiz, 74801'),
 ('Kwik E-mart', 'Calle de Georgina Berenguer 153 Puerta 4  Cádiz, 08223'),
 ('Kwik E-mart', 'Rambla de Raquel Zabaleta 49 Apt. 63  Cádiz, 49089'),
 ('Kwik E-mart', 'Callejón de Tomasa Saavedra 44 Apt. 45  Cádiz, 04059'),
 ('tío zetamart', 'Urbanización de Jose Manuel Serrano 19 Cádiz, 17607'),
 ('pc_faketory', 'Callejón Iker Quirós 184 Puerta 1  Cádiz, 94569'),
 ('pc_faketory', 'Cuesta Fermín Piña 57 Cádiz, 29999'),
 ('Kwik E-mart', 'Vial de Reinaldo Novoa 68 Puerta 0  Cádiz, 56894'),
 ('instatienda', 'Paseo Elvira Jiménez 60 Cádiz, 61533'),
 ('instatienda', 'Cuesta de Pili Sevilla 59 Cádiz, 80955'),
 ('tío zetamart', 'Acceso Imelda Tur 7 Cádiz, 69499'),
 ('Comic center', 'Via Ruth Burgos 29 Cádiz, 01194')]

**6.** Los nombres de productos que sean vendidos por la tienda `tío zetamart` y la tienda `pc_faketory`. Ten en cuenta que cuando decimos "productos que sean vendidos por ambas tiendas", son en realidad los productos que tengan el mismo nombre. **(0.4 pts)**

Pauta: 
- 0.2 por condición sobre nombre de tienda (así que había que hacer join de prouctos y tiendas, no servía buscar la tienda y poner el id no más)
- 0.2 intersectar infomación ya sea con INTERSECT o con algún otro join

In [ ]:
%%sql
SELECT Products.name
FROM Products, Stores
WHERE Products.store_id=Stores.id AND Stores.name = 'tío zetamart'
INTERSECT
SELECT Products.name
FROM Products, Stores
WHERE Products.store_id=Stores.id AND Stores.name = 'pc_faketory'

 * sqlite:///tu_shop_mini.db
Done.


[('Firewatch',), ('RAGE 2',)]

**7.** La cantidad de unidades vendidas de cada producto en cada tienda. **(0.4 pts)**


Pauta: 
- 0.1 hacer el join de todo correctamente 
- 0.2 por agrupar correctamente por id de tienda y de producto (tienen que se ambas, y ambas tienen que ser por id)
- 0.1 por hacer la agregación  para sacar las unidades vendidas

**OJO: aunque la consulta llegara a dar el resultado correcto, si no agrupan como mencioné, con ids, y por amabs cosas, descontar 0.2**

In [ ]:
%%sql
SELECT Stores.name AS store_name, Products.name AS product_name, SUM(BillVariants.quantity) AS units_sold
FROM Stores, Bills, Products, Variants, BillVariants
WHERE Stores.id = Products.store_id
AND Variants.product_id = Products.id
AND BillVariants.variant_id = Variants.id
AND BillVariants.bill_id = Bills.id
GROUP BY Stores.id, Products.id


 * sqlite:///tu_shop_mini.db
Done.


[('Comic center', 'The Legend of Zelda: Breath of the Wild', 1),
 ('Comic center', 'Civilization IV', 1),
 ('Comic center', 'Civilization IV', 1),
 ('Comic center', 'Day of Defeat', 2),
 ('Comic center', 'Pong', 2),
 ('Comic center', 'Civilization III', 2),
 ('Comic center', 'Amazing Cultivation Simulator / 了不起的修仙模拟器', 2),
 ('Comic center', 'Civilization III', 1),
 ('Comic center', 'Pokémon Ruby', 2),
 ('Comic center', 'Call of Duty: Black Ops III', 1),
 ('Comic center', 'Civilization IV', 3),
 ('Comic center', 'Portal', 1),
 ('Comic center', 'The Legend of Zelda: Majoras Mask', 3),
 ('Comic center', 'Pokémon Ruby', 2),
 ('Comic center', 'The Legend of Zelda: Skyward Sword', 1),
 ('Comic center', 'Dota 2', 7),
 ('Comic center', 'Portal 2', 4),
 ('Comic center', 'FTL: Faster Than Light', 5),
 ('Comic center', 'Super Mario Sunshine', 3),
 ('Comic center', 'Pokémon Ruby', 2),
 ('Comic center', 'The Legend of Zelda: Ocarina of Time', 2),
 ('Comic center', 'Civilization V', 1),
 ('Comic center', 'Commandos: Behind Enemy Lines', 1),
 ('Comic center', 'Pong', 4),
 ('Comic center', 'Castle Crashers', 1),
 ('Comic center', 'Battlefleet Gothic: Armada 2', 4),
 ('Comic center', 'Starcraft', 5),
 ('Comic center', 'Quake', 2),
 ('Comic center', 'American Truck Simulator', 6),
 ('Comic center', 'Pokémon White', 1),
 ('Comic center', 'Pokémon White 2', 1),
 ('Comic center', 'Life is Strange', 1),
 ('Comic center', 'Civilization IV', 9),
 ('Comic center', 'Civilization IV', 3),
 ('Comic center', 'Left 4 Dead 2', 1),
 ('Comic center', 'The Legend of Zelda: Breath of the Wild', 5),
 ('Comic center', 'Civilization IV', 3),
 ('Comic center', 'Pokémon FireRed', 2),
 ('Comic center', 'Pokémon Ruby', 1),
 ('Comic center', 'Metal Gear Solid', 10),
 ('Comic center', 'Pong', 3),
 ('Comic center', 'Banjo-Kazooie: Nuts & Bolts', 10),
 ('Comic center', 'Civilization IV', 1),
 ('Comic center', 'Call of Duty: World at War', 1),
 ('Comic center', 'The Legend of Zelda: Twilight Princess', 2),
 ('Comic center', 'Far Cry® Primal', 1),
 ('Comic center', 'Dota 2', 7),
 ('Comic center', 'Hobo: Tough Life', 4),
 ('Comic center', 'Pokémon Ruby', 2),
 ('Comic center', 'The Legend of Zelda: Majoras Mask', 4),
 ('Comic center', 'Pong', 5),
 ('Comic center', 'Crysis 2', 2),
 ('Comic center', 'Civilization IV', 7),
 ('Comic center', 'Pokémon Black', 6),
 ('Comic center', 'Max Payne 2: The Fall of Max Payne', 5),
 ('Comic center', 'Cuisine Royale', 13),
 ('Comic center', 'Life is Strange', 6),
 ('Comic center', 'F.E.A.R.', 3),
 ('Comic center', 'Halo Wars', 1),
 ('Comic center', 'Puyo Puyo Tetris', 2),
 ('Comic center', 'Katamari Damacy', 1),
 ('Comic center', 'Clannad', 4),
 ('Comic center', 'Super Smash Bros. Brawl', 2),
 ('Comic center', 'Bayonetta 2', 2),
 ('Comic center', 'The Wonderful 101', 8),
 ('Comic center', 'The Wonderful 101', 11),
 ('Comic center', 'Space Hulk: Tactics', 6),
 ('Comic center', 'The Legend of Zelda: Majoras Mask', 8),
 ('Comic center', 'Pokémon Ruby', 4),
 ('Comic center', 'Rust', 4),
 ('Comic center', 'Pokémon Black', 5),
 ('Comic center', 'Super Mario Bros.', 3),
 ('Comic center', 'Space Engineers', 9),
 ('Comic center', 'Football Manager 2019', 5),
 ('Comic center', 'Foundation', 5),
 ('Comic center', 'SoulWorker', 2),
 ('Comic center', 'Brütal Legend', 4),
 ('Comic center', 'Battletoads', 10),
 ('Comic center', 'Brothers: A Tale of Two Sons', 3),
 ('Comic center', 'Forager', 10),
 ('Comic center', 'SOULCALIBUR™ VI', 5),
 ('Comic center', 'Pokémon Black 2', 1),
 ('Comic center', 'Fire Emblem Awakening', 1),
 ('Comic center', 'Banjo-Kazooie: Nuts & Bolts', 7),
 ('Comic center', 'Super Mario Bros.', 5),
 ('Comic center', 'Banjo-Kazooie: Gruntys Revenge', 3),
 ('Comic center', 'Sonic Mania', 4),
 ('Comic center', 'Imperator: Rome', 8),
 ('Comic center', 'Kingdom Two Crowns', 2),
 ('Comic center', 'Hearts of Iron IV', 12),
 ('Comic center', 'Super Meat Boy', 11),
 ('Comic center', 'Civilization IV', 5),
 ('Comic center', 'Sonic Forces', 8),
 ('

**8.** El valor de las ventas de cada tienda en marzo (en $$$), ordenadas por ventas. **(0.4 pts)** **Pro Tip:** Para dividir dos ints y que el resultado sea float hay que convertir uno de los ints a float primero, lo que se puede hacer multiplicando por `1.0`. Algo así: `columna_a*1.0/100`

Pauta: 
- 0.2 por la agregación. Debían hacer ellos el descuento, si no lo hacen, descontar 0.1.
- 0.1 por agrupar por id de tienda
- 0.1 por ordenar (ojalá ascendente,pero da lo mismo si lo hacen desc)



**OJO con el group by, tiene que ser por id, sino descuentan el 0.1**

In [12]:
%%sql
SELECT Stores.name, SUM(BillVariants.quantity * BillVariants.price * ((100 - BillVariants.discount_rate)*1.0/100)) AS sales
FROM Stores, Bills, Products, Variants, BillVariants
WHERE Stores.id = Products.store_id
AND Variants.product_id = Products.id
AND BillVariants.variant_id = Variants.id
AND BillVariants.bill_id = Bills.id
AND Bills.created_at BETWEEN '2021-03-01' AND '2021-04-01'
GROUP BY Stores.id
ORDER BY sales

 * sqlite:///tu_shop_mini.db
Done.


[('Comic center', 459680.0),
 ('Kwik E-mart', 905755.0),
 ('pc_faketory', 1578680.0),
 ('tío zetamart', 2475839.5),
 ('instatienda', 5018800.0)]

**9.** Los ids de las variantes que no han tenido ninguna venta en marzo y tienen stock disponible. **(0.4 pts)**

Pauta: 
- 0.2 por sacar las variantes que han tenido venta en marzo
- 0.2 de alguna manera hacer la resta

In [ ]:
%%sql
SELECT Variants.id AS id
FROM VARIANTS
WHERE Variants.quantity > 0
EXCEPT
SELECT Variants.id
FROM Variants JOIN BillVariants
ON Variants.id = BillVariants.variant_id
WHERE BillVariants.created_at BETWEEN '2021-03-01' AND '2021-04-01'

 * sqlite:///tu_shop_mini.db
Done.


[(1,),
 (3,),
 (13,),
 (28,),
 (31,),
 (36,),
 (37,),
 (45,),
 (53,),
 (59,),
 (61,),
 (68,),
 (69,),
 (70,),
 (71,),
 (72,),
 (73,),
 (74,),
 (75,),
 (78,),
 (79,),
 (80,),
 (81,),
 (82,),
 (83,),
 (84,),
 (86,),
 (87,),
 (88,),
 (89,),
 (94,),
 (95,),
 (96,),
 (97,),
 (98,),
 (99,),
 (100,),
 (101,),
 (104,),
 (105,),
 (827,),
 (831,),
 (832,),
 (833,),
 (834,),
 (835,),
 (899,),
 (900,),
 (901,),
 (902,),
 (911,),
 (912,),
 (913,),
 (914,),
 (915,),
 (916,),
 (917,),
 (918,),
 (919,),
 (920,),
 (921,),
 (922,),
 (923,),
 (924,),
 (925,),
 (926,),
 (927,),
 (928,),
 (929,),
 (930,),
 (931,),
 (932,),
 (933,),
 (934,),
 (935,),
 (936,),
 (937,),
 (938,),
 (939,),
 (940,),
 (941,),
 (942,),
 (943,),
 (944,),
 (945,),
 (946,),
 (947,),
 (948,),
 (949,),
 (950,),
 (951,),
 (952,),
 (953,),
 (954,),
 (955,),
 (956,),
 (957,),
 (958,),
 (960,),
 (961,),
 (962,),
 (1074,),
 (1075,),
 (1076,),
 (1077,),
 (1078,),
 (1081,),
 (1082,),
 (1083,),
 (1446,),
 (1447,),
 (1451,),
 (1454,),
 (1456,),
 (1459,),
 (1462,),
 (1465,),
 (1468,),
 (1471,),
 (1474,),
 (1476,),
 (1477,),
 (1478,),
 (1479,),
 (1480,),
 (1483,),
 (1486,),
 (1487,),
 (1488,),
 (1489,),
 (1490,),
 (1492,),
 (1495,),
 (1496,),
 (1498,),
 (1499,),
 (1500,),
 (1501,),
 (1505,),
 (1516,),
 (1518,),
 (1522,),
 (1523,),
 (1725,),
 (1727,),
 (1728,),
 (1729,),
 (1731,),
 (1732,),
 (1733,),
 (1734,),
 (1735,),
 (1737,),
 (1739,),
 (1740,),
 (1741,),
 (1742,),
 (1743,),
 (1744,),
 (1745,),
 (1746,),
 (1747,),
 (1959,),
 (2109,),
 (2110,),
 (2111,),
 (2112,),
 (2113,),
 (2114,),
 (2120,),
 (2207,),
 (2208,),
 (2210,),
 (2212,),
 (2214,),
 (2217,),
 (2218,),
 (2220,),
 (2221,),
 (2222,),
 (2223,),
 (2224,),
 (2226,),
 (2245,),
 (2250,),
 (2255,),
 (2256,),
 (2260,),
 (2575,),
 (2598,),
 (2604,),
 (2627,),
 (2628,),
 (2700,),
 (2702,),
 (2705,),
 (2707,),
 (2712,),
 (2713,),
 (2714,),
 (2740,),
 (2742,),
 (2744,),
 (2746,),
 (2765,),
 (2767,),
 (2768,),
 (2771,),
 (2773,),
 (2775,),
 (2776,),
 (2856,),
 (2906,),
 (3064,),
 (3067,),
 (3069,),
 (3070,),
 (3071,),
 (3072,),
 (3073,),
 (3074,),
 (3076,),
 (3077,),
 (3078,),
 (3079,),
 (3080,),
 (3081,),
 (3085,),
 (3199,),
 (3202,),
 (3203,),
 (3207,),
 (3209,),
 (3237,),
 (3238,),
 (3249,),
 (3250,),
 (3251,),
 (3253,),
 (3287,),
 (3322,),
 (3331,),
 (3332,),
 (3334,),
 (3336,),
 (3337,),
 (3338,),
 (3339,),
 (3340,),
 (3341,),
 (3363,),
 (3549,),
 (3552,),
 (3553,),
 (3556,),
 (3561,),
 (3562,),
 (3563,),
 (3564,),
 (3565,),
 (3566,),
 (3567,),
 (3568,),
 (3569,),
 (3832,),
 (3833,),
 (3865,),
 (3868,),
 (3871,),
 (3897,),
 (3916,),
 (3917,),
 (3921,),
 (3922,),
 (3923,),
 (3924,),
 (3925,),
 (3929,),
 (3931,),
 (3932,),
 (3935,),
 (3936,),
 (3937,),
 (3938,),
 (3941,),
 (3942,),
 (3943,),
 (3944,),
 (3945,),
 (3946,),
 (3947,),
 (3948,),
 (3950,),
 (3951,),
 (3952,),
 (3953,),
 (3954,),
 (3964,),
 (4074,),
 (4075,),
 (4078,),
 (4082,),
 (4089,),
 (4093,),
 (4094,),
 (4095,),
 (4096,),
 (4097,),
 (4099,),
 (4105,),
 (4107,),
 (4158,),
 (4224,),
 (4270,),
 (6629,),
 (7858,),
 (7860,),
 (7861,),
 (7862,),
 (7863,),
 (7864,),
 (7866,),
 (7867,),
 (7868,),
 (7869,),
 (8714,),
 (9975,),
 (9976,),
 (9979,),
 (9980,),
 (9981,),
 (9982,),
 (9983,),
 (9984,),
 (9985,),
 (10171,),
 (12389,),
 (12950,),
 (13453,),
 (13455,),
 (13456,),
 (13457,),
 (13459,),
 (13461,),
 (13462,),
 (15148,),
 (15150,),
 (15413,),
 (15414,),
 (15415,),
 (15731,),
 (15732,),
 (15734,),
 (15736,),
 (15738,),
 (15739,),
 (15740,),
 (15741,),
 (15743,),
 (15744,),
 (15745,),
 (15746,),
 (15747,),
 (15751,),
 (15752,),
 (15753,),
 (15754,),
 (15755,),
 (15756,),
 (15757,),
 (15758,),
 (15762,),
 (15763,),
 (15765,),
 (15766,),
 (15767,),
 (15768,),
 (15769,),
 (15770,),
 (15771,),
 (15772,),
 (15776,),
 (17976,),
 (17977,),
 (17987,),
 (18134,),
 (18483,),
 (18925,),
 (19442,),
 (19443,),
 (19446,),
 (19447,),
 (19452,),
 (19455,),
 (19456,),
 (19457,),
 (19461,),
 (19462,),
 (19467,),
 (19472,),
 (19477,),
 (19482,),
 (19485,),
 (19487,),
 (19491,),
 (20506,),
 (20508,

**10.** El nombre y precio de la variante con el precio más caro. **(0.4 pts)**

Pauta: 
- 0.2 por sacar el más caro.
- 0.2 por obtener los nombres completos (producto + sufijo) variantes más caras

 Acá igual lo podían hacer con limit (onda ordenar y quedarse con el primero)
 
 Ojo que alguien puede haber pensado que para un producto, obtener su variante más cara. Si lo hizo así, igual considerarlo bueno.

In [ ]:
%%sql
SELECT (Products.name || ' ' || Variants.name_suffix) AS full_name, Variants.price
FROM Variants, Products
WHERE Variants.product_id = Products.id
AND Variants.price = (SELECT MAX(Variants.price) FROM Variants) 

 * sqlite:///tu_shop_mini.db
Done.


[('Assassin’s Creed® IV Black Flag™ ', 62900)]

De las siguientes 4 consultas solo debes hacer 3. Si haces las 4 no te revisaremos la última.

**11.** El valor promedio de venta de cada tienda en marzo (con sus respectivos nombres o ids). **(0.5 pts)**

Pauta: 
- 0.2 por la agregación de promedio (había que agrupar tiendas por id! si no lo hizo, descontar 0.1).
- 0.2 por agregación para obtener las ventas de una tienda (agrupar por id de tienda y el id de bills)
- 0.1 por intervalo de fechas correcta (marzo 2021)

**OJO con los group by**

In [10]:
%%sql
SELECT Stores.name, AVG(bill_sales.sales) AS average
FROM Stores JOIN (
    SELECT Stores.id AS store_id, SUM(BillVariants.quantity * BillVariants.price * ((100 - BillVariants.discount_rate)*1.0/100)) AS sales
    FROM Stores, Bills, Products, Variants, BillVariants
    WHERE Stores.id = Products.store_id
    AND Variants.product_id = Products.id
    AND BillVariants.variant_id = Variants.id
    AND BillVariants.bill_id = Bills.id
    AND Bills.created_at BETWEEN '2021-03-01' AND '2021-04-01'
    GROUP BY Stores.id, Bills.id) AS bill_sales
ON Stores.id = bill_sales.store_id
GROUP BY Stores.id

 * sqlite:///tu_shop_mini.db
Done.


[('Comic center', 27040.0),
 ('tío zetamart', 30947.99375),
 ('Kwik E-mart', 20585.340909090908),
 ('instatienda', 61960.49382716049),
 ('pc_faketory', 25462.58064516129)]

**12.** El porcentaje del total de clientes que fueron recurrentes en el mes de abril (recurrentes son aquellos que ya habían comprado alguna vez antes de comprar en abril). Los clientes se identifican por el email **(0.5 pts)**

Pauta: 
- 0.3 por calcular bien las recurrencia
- 0.2 por el resto

In [ ]:
%%sql
SELECT COUNT(DISTINCT old_clients.email)*1.0/COUNT(DISTINCT Bills.email) * 100 AS recurrence
FROM Bills LEFT OUTER JOIN (
    SELECT Bills.email AS email
    FROM Bills
    WHERE Bills.created_at < '2021-04-01'
    ) AS old_clients
ON Bills.email = old_clients.email
WHERE Bills.created_at > '2021-04-01'

 * sqlite:///tu_shop_mini.db
Done.


[(20.27027027027027,)]

**13.**  La cantidad de dias en promedio que han pasado desde que se creo cada venta, agrupado por status. **(0.5 pts)**

**Pro Tips**:
- `DATE(SUBSTR(columna, 0, 11))` Retorna un Date de sqlite a partir de una columna de fecha de nuesta base de datos.
-`DATE('now')` Retorna un objeto Date de sqlite con la fecha de hoy
-La función [JULIANDAY(date)](https://www.techonthenet.com/sqlite/functions/julianday.php#:~:text=The%20SQLite%20julianday%20function%20takes,as%20a%20floating%20point%20number.) Retorna los días que han pasado entre date y una fecha constante muy en el pasado.

Pauta: 
- 0.4 por calcular bien la cantidad de días promedio
- 0.1 agrupar por status

In [9]:
%%sql
SELECT Bills.status, AVG(JULIANDAY(DATE('now')) - JULIANDAY(DATE(SUBSTR(Bills.created_at, 0, 11)))) AS average_diff
FROM Bills
GROUP BY Bills.status

 * sqlite:///tu_shop_mini.db
Done.


[('awaiting_payment', 150.97122302158274),
 ('deleted', 174.39940828402368),
 ('payment_accepted', 134.40826873385012),
 ('payment_sent', 126.0),
 ('received', 147.24029574861368),
 ('shipped', 166.90401567091087)]

**14.** El nombre de la tienda que más ventas hizo en abril, junto con la cantidad de ventas. **Pro Tip**: La sentencia [LIMIT](https://www.sqlitetutorial.net/sqlite-limit/) que reduce la cantidad de resultados podría servir.

Pauta: 
- 0.2 por la condición de agregación
- 0.1 por agrupar por id de tienda y no otro atributo
- 0.2 por el resto (fechas, joins correctos...etc)

Podían hacerlo con o sin limit

**OJO con el group by, tiene que ser por id, sino descuentan el 0.1**

Sin `LIMIT`

In [ ]:
%%sql
SELECT  Stores.name, COUNT(bills.id) as sales_count
FROM Stores JOIN Bills
ON Stores.id = Bills.store_id
WHERE Bills.created_at BETWEEN '2021-04-01' AND '2021-05-01'
GROUP BY Stores.id
HAVING sales_count = (
  SELECT MAX(aux.total_bills)
  FROM (
    SELECT COUNT(Bills.id) as total_bills
    FROM Stores JOIN Bills
    ON Stores.id = Bills.store_id
    WHERE Bills.created_at BETWEEN '2021-04-01' AND '2021-05-01'
    GROUP BY Stores.id
  ) as aux
)

 * sqlite:///tu_shop_mini.db
Done.


[('Kwik E-mart', 80)]

Con `LIMIT`

In [11]:
%%sql
SELECT  Stores.name, COUNT(bills.id) as sales_count
FROM Stores JOIN Bills
ON Stores.id = Bills.store_id
WHERE Bills.created_at BETWEEN '2021-04-01' AND '2021-05-01'
GROUP BY Stores.id
ORDER BY sales_count DESC
LIMIT 1

 * sqlite:///tu_shop_mini.db
Done.


[('Kwik E-mart', 80)]

## Parte 2: Índices (1 punto, 0.1 cada respuesta) 

En esta parte del control tienen qué responder cual será el costo I/O para ejecutar consultas SQL ocupando distintos tipos de índices.

En particular, en este ejercicio asumiremos qué tenemos una relación Alumno(id int PRIMARY KEY, nombre varchar(100), promedio float).

A nuestra disposición encontramos dos índices: 
- Un B+tree *clustered* sobre id, y 
- Un hash index *unclustered* sobre el promedio.

La relación tiene 320,000 tuplas. En cada página del disco nos cabe 100 tuplas.El promedio toma valores entre 3.9 y 7.0, en incrementos de 0.1 (i.e. hay solo 32 posibles promedio). Los ids van desde 1 a 320,000.

El árbol es de altura *h*, y la hojas están ocupadas a 80%. 

En el Hash Index cada página contiene 1,000 punteros a las páginas del disco, y hay 32 buckets (uno para cada promedio). Las notas son distribuidas de manera uniforme y **si hay overflow pages** (deberían calcular cuanto).

¿Cuál es el costo de evaluar las siguientes consultas con los dos índices (en el peor de los casos)?
1. SELECT * FROM Alumno WHERE id=12
2. SELECT * FROM Alumno WHERE nombre = 'Adrián Soto'
3. SELECT * FROM Alumno WHERE promedio = 5.6
4. SELECT * FROM Alumno WHERE id > 1000 AND id <= 2000
5. SELECT * FROM Alumno WHERE promedio <4.9

## Solución

En una página del disco caben 100 tuplas.

La relación entera tiene 3,200 páginas en el disco.

Cada bucket del Hash Index tiene 9 overflow pages (y una principal). En total hay 32 x 10 x 1000 = 320,000 tuplas.

| Consulta| B+ tree | Hash Index |
| ------- | -------- | ------------ |
| 1       | h | 320 (índice) + 320.000 (si lo hacemos a la mala), o 320 + 3,200 (si chequeamos si ya tuvimos la página) |
| 2       | (h - 1) + 3,200/0.8 | 320 (índice) + 320.000 |
| 3    | (h - 1) + 3,200/0.8 | 10 + 10.000 |
| 4    | h + 10/0.8 | 320 (índice) + 320.000 |
| 5    | (h - 1) + 3,200/0.8 | 10 * 10 + 10 * 10 * 1,000|

# Pauta

Simplemente 0.1 pto por cada valor correcto. 

Notar que la idea es que los resultados consideren el -1 (por ejemplo h-1), es decir, que es h-1 hasta antes de llegar al final del árbol, y despúes se van sumando las páginas que se recorren. Pero si alguien consideró h hasta antes de llegar al final (entonces donde dice h-1 consideran solo h o cosas así), no descuenten pero dejen el comentario de que debería considerarse. 